python/lib/core/dmod/core/context.py

```diff
@@ -0,0 +1,607 @@
+        ...
+
+
+@typing.runtime_checkable
+class ProxiableDeletePropertyProtocol(typing.Protocol):
+    """
+    Outline for a class that can explicitly delete a property value
+    """
+    def __del_property__(self, key: str) -> None:
+        ...
+
+
+class ProxiablePropertyMixin(ProxiableGetPropertyProtocol, ProxiableSetPropertyProtocol, ProxiableDeletePropertyProtocol):
+    """
+    Mixin functions that allow property functions (fget, fset, fdel) to be called explicitly rather than implicitly
+    """
+    def __get_property__(self, key: str) -> typing.Any:
+        field = getattr(self.__class__, key)
+        if not isinstance(field, property):
+            raise TypeError(f"'{key}' is not a property of type '{self.__class__.__name__}'")
+
+        if field.fget is None:
+            raise Exception(f"Cannot retrieve the value for '{key}' on type '{self.__class__.__name__} - it is write-only")
+        return field.fget(self)
+
+    def __set_property__(self, key: str, value) -> None:
+        field = getattr(self.__class__, key)
+        if not isinstance(field, property):
+            raise TypeError(f"'{key}' is not a property of type '{self.__class__.__name__}'")
+
+        if field.fset is None:
+            raise Exception(f"Cannot modify '{key}' on type '{self.__class__.__name__}' - it is read-only")
+
+        field.fset(self, value)
+
+    def __del_property__(self, key: str) -> None:
+        field = getattr(self.__class__, key)
+        if not isinstance(field, property):
+            raise TypeError(f"'{key}' is not a property of type '{self.__class__.__name__}'")
+
+        if field.fdel is None:
+            raise Exception(f"The property '{key}' cannot be deleted from a type '{self.__class__.__name__}'")
+
+        field.fdel(self)
+
+
+def is_property(obj: object, member_name: str) -> bool:
+    """
+    Checks to see if a member of an object is a property
+
+    Args:
+        obj: The object to check
+        member_name: The member on the object to check
+
+    Returns:
+        True if the member with the given name on the given object is a property
+    """
+    if not hasattr(obj, member_name):
+        raise AttributeError(f"{obj} has no attribute '{member_name}'")
+
+    if isinstance(obj, type):
+        return isinstance(getattr(obj, member_name), property)
+
+    # Is descriptor: inspect.isdatadescriptor(dict(inspect.getmembers(obj.__class__))[member_name])
+    parent_reference = dict(inspect.getmembers(obj.__class__))[member_name]
+    return isinstance(parent_reference, property)
+
+
+def form_proxy_name(cls: type) -> str:
+    """
+    Programmatically form a name for a proxy class
+
+    Args:
+        cls: The class that will end up with a proxy
+    Returns:
+        The accepted name for a proxy class
+    """
+    if not hasattr(cls, "__name__"):
+        raise TypeError(f"Cannot create a proxy name for {cls} - it has no consistent '__name__' attribute")
+
+    return f"{cls.__name__}{PROXY_SUFFIX}"
```

Comment:
`__qualname__` provides the full name of the class starting from its root package, so something like `dmod.core.common.collections.Whatsit` instead of `Whatsit`. This function is supposed to get the name of a class, so naming something `dmod.core.common.collections.WhatsitProxy` instead of `WhatsitProxy` will yield a syntax error.

Reply:
> `__qualname__` provides the full name of the class starting from its root package, so something like `dmod.core.common.collections.Whatsit` instead of `Whatsit`.

That's not true. From the [pep](https://peps.python.org/pep-3155/#proposal)

> For top-level functions and classes, the `__qualname__` attribute is equal to the `__name__` attribute. For nested classes, methods, and nested functions, the `__qualname__` attribute contains a dotted path leading to the object from the module top-level. A function’s local namespace is represented in that dotted path by a component named <locals>.

Reply:
Ok, I'm going to experiment with it to ensure that everything's kosher.  I feel like I ran afoul of this when I was first writing it.